File: Protocol/Responses/SendMessageResponse.cpp
#include "SendMessageResponse.h"

SendMessageResponse::SendMessageResponse(SendMessageResult result)
{
    this->result = result;
}

SendMessageResult SendMessageResponse::GetResult() const
{
    return this->result;
}

ResponseType SendMessageResponse::GetResponseType() const
{
    return ResponseType::SendMessage;
}

string SendMessageResponse::ToString() const
{
    ostringstream stream;
    stream << (int)GetResponseType() << ' ' << (int)GetResult();
    return stream.str();
}

SendMessageResponse *SendMessageResponse::Parse(string str)
{
    istringstream stream(str);
    int responseType, result;
    stream >> responseType >> result;

    if ((ResponseType)responseType != ResponseType::SendMessage)
        throw "Invalid type of request";

    return new SendMessageResponse((SendMessageResult)result);
}

File: Client/Protocol/Responses/SendMessageResponse.cs
using System;
using System.IO;
using System.Text;

namespace NetTalk.Protocol.Responses
{
    public enum SendMessageResult
    {
        Ok,
        UserNotFound,
    }

    public class SendMessageResponse : ResponseBase
    {
        public SendMessageResponse(SendMessageResult result)
        {
            Result = result;
        }

        public override ResponseType ResponseType { get; } = ResponseType.SendMessage;

        public SendMessageResult Result { get; private set; }

        public static SendMessageResponse Parse(byte[] bytes)
        {
            using (MemoryStream stream = new MemoryStream(bytes))
            using (BinaryReader reader = new BinaryReader(stream, Encoding.ASCII))
            {
                ResponseType responseType = (ResponseType)reader.ReadInt32();
                if (responseType != ResponseType.SendMessage)
                    throw new InvalidOperationException("Invalid type of response");

                SendMessageResult result = (SendMessageResult)reader.ReadInt32();
                return new SendMessageResponse(result);
            }
        }
    }
}